packages/@ionic/cli/src/lib/oauth/oauth.ts

```diff
@@ -22,53 +24,74 @@ export interface TokenParameters {
 }
 
 export interface OAuth2FlowOptions {
-  readonly authorizationUrl: string;
-  readonly tokenUrl: string;
-  readonly clientId: string;
   readonly redirectHost?: string;
   readonly redirectPort?: number;
+  readonly accessTokenRequestContentType?: ContentType;
 }
 
 export interface OAuth2FlowDeps {
   readonly client: IClient;
+  readonly config: IConfig;
 }
 
-export abstract class OAuth2Flow {
-  readonly authorizationUrl: string;
-  readonly tokenUrl: string;
-  readonly clientId: string;
+export abstract class OAuth2Flow<T> {
+  abstract readonly flowName: string;
+  readonly oauthConfig: OAuthServerConfig;
   readonly redirectHost: string;
   readonly redirectPort: number;
+  readonly accessTokenRequestContentType: ContentType;
 
-  constructor({ authorizationUrl, tokenUrl, clientId, redirectHost = REDIRECT_HOST, redirectPort = REDIRECT_PORT }: OAuth2FlowOptions, readonly e: OAuth2FlowDeps) {
-    this.authorizationUrl = authorizationUrl;
-    this.tokenUrl = tokenUrl;
-    this.clientId = clientId;
+  constructor({ redirectHost = REDIRECT_HOST, redirectPort = REDIRECT_PORT, accessTokenRequestContentType = ContentType.JSON }: OAuth2FlowOptions, readonly e: OAuth2FlowDeps) {
+    this.oauthConfig = this.getAuthConfig();
     this.redirectHost = redirectHost;
     this.redirectPort = redirectPort;
+    this.accessTokenRequestContentType = accessTokenRequestContentType;
   }
 
   get redirectUrl(): string {
     return `http://${this.redirectHost}:${this.redirectPort}`;
   }
 
-  async run(): Promise<string> {
+  async run(): Promise<T> {
     const verifier = this.generateVerifier();
     const challenge = this.generateChallenge(verifier);
 
     const authorizationParams = this.generateAuthorizationParameters(challenge);
-    const authorizationUrl = `${this.authorizationUrl}?${qs.stringify(authorizationParams)}`;
+    const authorizationUrl = `${this.oauthConfig.authorizationUrl}?${qs.stringify(authorizationParams)}`;
 
     await openUrl(authorizationUrl);
 
     const authorizationCode = await this.getAuthorizationCode();
-    const token = await this.getAccessToken(authorizationCode, verifier);
+    const token = await this.exchangeAuthForAccessToken(authorizationCode, verifier);
 
     return token;
   }
 
+  async exchangeRefreshToken(refreshToken: string): Promise<T> {
+    const params = this.generateRefreshTokenParameters(refreshToken);
+    const { req } = await this.e.client.make('POST', this.oauthConfig.tokenUrl, this.accessTokenRequestContentType);
+
+    const res = await req.send(params);
+
+    // check the response status code first here
+    if (!res || !res.body || !res.status || res.status < 200 || res.status >= 300) {
```

Comment:
I think you can use `!res.ok`: https://visionmedia.github.io/superagent/#response-status